source/validate_cfg.cpp

```diff
@@ -24,32 +24,303 @@
+        staged.emplace_back(block_info{child, begin(child->get_successors())});
+        processed.insert(child->get_id());
+      }
+      top.iter++;
+    }
+  }
+  return out;
+}
+}
+
+vector<pair<BasicBlock*, BasicBlock*>> CalculateDominators(
+    const BasicBlock& first_block) {
+  struct block_detail {
+    size_t dominator;
+    size_t postorder_index;
+  };
+
+  vector<cbb_ptr> postorder = PostOrderSort(first_block, 10);
+  const size_t undefined_dom = static_cast<size_t>(postorder.size());
+
+  unordered_map<cbb_ptr, block_detail> idoms;
+  for (size_t i = 0; i < postorder.size(); i++) {
+    idoms[postorder[i]] = {undefined_dom, i};
+  }
+
+  idoms[postorder.back()].dominator = idoms[postorder.back()].postorder_index;
+
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    for (auto b = postorder.rbegin() + 1; b != postorder.rend(); b++) {
+      size_t& b_dom = idoms[*b].dominator;
+      const vector<BasicBlock*>& predecessors = (*b)->get_predecessors();
+
+      // first processed predecessor
+      auto res = find_if(begin(predecessors), end(predecessors),
+                         [&idoms, undefined_dom](BasicBlock* pred) {
+                           return idoms[pred].dominator != undefined_dom;
+                         });
+      assert(res != end(predecessors));
+      BasicBlock* idom = *res;
+      size_t idom_idx = idoms[idom].postorder_index;
+
+      // all other predecessors
+      for (auto p : predecessors) {
+        if (idom == p || p->is_reachable() == false) {
+          continue;
+        }
+        if (idoms[p].dominator != undefined_dom) {
+          size_t finger1 = idoms[p].postorder_index;
+          size_t finger2 = idom_idx;
+          while (finger1 != finger2) {
+            while (finger1 < finger2) {
+              finger1 = idoms[postorder[finger1]].dominator;
+            }
+            while (finger2 < finger1) {
+              finger2 = idoms[postorder[finger2]].dominator;
+            }
+          }
+          idom_idx = finger1;
+        }
+      }
+      if (b_dom != idom_idx) {
+        b_dom = idom_idx;
+        changed = true;
+      }
+    }
+  }
+
+  vector<pair<bb_ptr, bb_ptr>> out;
+  for (auto idom : idoms) {
+    // NOTE: performing a const cast for convenient usage with
+    // UpdateImmediateDominators
+    out.push_back({const_cast<BasicBlock*>(get<0>(idom)),
+                   const_cast<BasicBlock*>(postorder[get<1>(idom).dominator])});
+  }
+  return out;
+}
+
+void UpdateImmediateDominators(vector<pair<bb_ptr, bb_ptr>>& dom_edges) {
+  for (auto& edge : dom_edges) {
+    get<0>(edge)->SetImmediateDominator(get<1>(edge));
+  }
+}
+
+void printDominatorList(BasicBlock& b) {
+  std::cout << b.get_id() << " is dominated by: ";
+  const BasicBlock* bb = &b;
+  while (bb->GetImmediateDominator() != bb) {
+    bb = bb->GetImmediateDominator();
+    std::cout << bb->get_id() << " ";
+  }
+}
+
+#define CFG_ASSERT(ASSERT_FUNC, TARGET) \
```

Comment:
this is in a cpp file. I will probably use this throughout the file so the undef will most likely be on the last line. I can always expand the macro but I though this was easier to read.